mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,708 @@
+        elif action == AnnotatorAction.UNANNOTATE.value:
+            rules = self.config[AnnotatorAction.UNANNOTATE.value]
+
+        for rule in rules:
+            pattern = Annotator._get_pattern(rule)
+            filtered_paths = self._resource_loader.filter_file_paths(
+                file_pattern=pattern, file_paths=all_file_paths
+            )
+            for path in filtered_paths:
+                entities = self._get_entities(rule)
+                file_entities_map[path] = entities
+        return file_entities_map
+
+    @staticmethod
+    def _get_pattern(rule):
+        """ Convert a rule represented as a dictionary with the keys "domains", "intents",
+        "entities" into a regex pattern.
+
+        Args:
+            rule (dict): Annotation/Unannotation rule.
+
+        Returns:
+            pattern (str): Regex pattern specifying allowed file paths.
+        """
+        pattern = []
+        for x in ["domains", "intents", "files"]:
+            processed_segment = Annotator._process_segment(rule[x])
+            pattern.append(processed_segment)
+        pattern = "/".join(pattern)
+        return ".*/" + pattern
+
+    @staticmethod
+    def _process_segment(segment):
+        """ Process an individual segment from a rule dictionary.
+
+        Args:
+            segment (str): Section of a rule dictionary ("domains", "intents", "entities").
+
+        Returns:
+            segment (str): Cleaned section of the rule dictionary.
+        """
+        segment = re.sub("[()]", "", segment)
+        segment = segment.replace(".*", ".+")
+        segment = segment.replace("*", ".+")
+        segment = segment.split("|")
+        segment = "|".join([x.strip() for x in segment])
+        segment = "(" + segment + ")" if "|" in segment else segment
+        return segment
+
+    def _get_entities(self, rule):
+        """ Process the entities specified in a rule dictionary. Check if they are valid
+        for the given annotator.
+
+        Args:
+            rule (dict): Annotation/Unannotation rule with an "entities" key.
+
+        Returns:
+            valid_entities (list): List of valid entities specified in the rule.
+        """
+        if rule["entities"].strip() == "*":
+            return ["*"]
+        entities = re.sub("[()]", "", rule["entities"]).split("|")
+        valid_entities = []
+        for entity in entities:
+            entity = entity.strip()
+            if self.valid_entity_check(entity):
+                valid_entities.append(entity)
+            else:
+                logger.warning("%s is not a valid entity. Skipping entity.", entity)
+        return valid_entities
+
+    @abstractmethod
+    def valid_entity_check(self, entity):
+        """ Determine if an entity type is valid.
+
+        Args:
+            entity (str): Name of entity to annotate.
+
+        Returns:
+            bool: Whether entity is valid.
+        """
+        return True
```

Comment:
Is this function needed?

Reply:
The abstract method valid_entity_check() returns True by default but developers are encouraged to overwrite this function in their implementation class to only support a custom list of entities.

Should I raise a NotImplementedException instead?